sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java

```diff
@@ -426,29 +426,40 @@ public static Boolean checkFileExistsAnySymlinks(Path path, boolean neverFollowS
                 // this is a bad fix because this leaves a nasty race condition - the directory may turn into a symlink
                 // between this check and the call to open()
                 for (int i = 1; i <= path.getNameCount(); i++) {
-                    Path checkForSymLink = getPartialPath(path, i);
+                    Path checkForSymLink = getFirstPartsOfPath(path, i);
                     BasicFileAttributes basicFileAttributes
                             = Files.readAttributes(checkForSymLink, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
                     if (basicFileAttributes.isSymbolicLink()) {
                         return false;
                     }
                 }
             }
-            return Boolean.TRUE;
+            return true;
         } catch (NoSuchFileException e) {
-            return Boolean.FALSE;
+            return false;
         } catch (IOException e) {
             return null;
         }
     }
 
-    private static Path getPartialPath(Path path, int partsToExtract) {
+    /**
+     * Extracts the first n parts of the path. For example <br>
+     * ("/home/test/test12", 1) returns "/home", <br>
+     * ("/home/test", 1) returns "/home/test" <br>
+     * etc.
+     *
+     * @param  path           the path to extract parts of
+     * @param  partsToExtract the number of parts to extract
+     * @return                the extracted path
+     */
+    public static Path getFirstPartsOfPath(Path path, int partsToExtract) {
         String firstName = path.getName(0).toString();
         String[] names = new String[partsToExtract - 1];
         for (int j = 1; j < partsToExtract; j++) {
             names[j - 1] = path.getName(j).toString();
         }
         Path checkForSymLink = path.getFileSystem().getPath(firstName, names);
+        // the root is not counted as a directory part so we must resolve the result relative to it.
         if (path.getRoot() != null) {
             checkForSymLink = path.getRoot().resolve(checkForSymLink);
         }
```

Comment:
> in order for this to be optimized there would have to be lots of calls to the code

If there aren't many calls to the code, its performance is irrelevant. This is why the Oracle JVM JIT compiler engine is called hotspot - it optimizes only the spots that matter.

> the JVM cannot be sure that the return value of getRoot is the same

Read the implementation of [UnixPath](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/unix/classes/sun/nio/fs/UnixPath.java). Hotspot is smart enough to see the code path in `getRoot()` does not for the same Path object, and that it always returns the same root from [UnixFileSystem](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java#L92).

In short, there's no practical performance concern. At best, in theory, this might save a few microseconds (more likely nanoseconds) on application startup. And in the words of Donald Knuth, "Premature optimization is the root of all evil." Mentioning performance in irrelevant places is bad practice.

> Path is definitely not immutable

If you have a mutable Path implementation, you probably shouldn't be using standard libraries with it. And you definitely shouldn't be mutating it while passing it into a library function.

In this case, a `Path` implementation that breaks this code is a theoretical exercise. It would require `getRoot()` to mutate its state, or another thread to cause a race condition in `getRoot()`.

Reply:
> readability and especially debugging concerns

I don't agree with this either, but as I said, it's a matter of style. Repeating the call saves a line of code and a local variable. It's more concise, and easier for me to read. Modern Java debuggers can evaluate an expression at a breakpoint, so I wouldn't find this any easier to debug. Programmer discretion is required, in these sorts of questions, and we came to different conclusions. I understand the your view and think that the extra variable is also acceptable, but still prefer the other approach.

> Who says the Path being examined is a UnixPath or WindowsPath ?

It's reasonable to assume that a simple function like `getRoot()` on a `Path` object runs fast. Remember that the Path object is a representation of a file in a file file system, and the methods on `Path` (other than `toRealPath()``) are done without IO. This is why they cannot throw `IOException`. These operations thus are fast by their nature, and Hotspot does a great job optimizing them.

>  What guarantee do you have that implementation will not change to something more "expensive" ?

In addition to the above analysis, historically the JVM does not make changes that significantly slow performance. I'm happy to bet on this in places where I think it's cleaner. I would agree with this concern if the function `getRoot()` was non-trivial.